lib/sycamore/sycamore/transforms/extract_graph.py

```diff
@@ -0,0 +1,155 @@
+from abc import ABC, abstractmethod
+from typing import TYPE_CHECKING, Dict, Any
+from sycamore.plan_nodes import Node
+from sycamore.transforms.map import Map
+from sycamore.data import Document, MetadataDocument
+import uuid
+
+if TYPE_CHECKING:
+    from sycamore.docset import DocSet
+
+
+class GraphData(ABC):
+    def __init__(self):
+        pass
+
+
+class GraphMetadata(GraphData):
+    """
+    Object which handles what fields to extract metadata from and what fields to represent them as in neo4j
+
+    Args:
+        nodeKey: Key used to access document metadata in the document['properties] dictionary
+        nodeLabel: The label used in neo4j the node of a piece of metadata
+        relLabel: The label used in neo4j for the relationship between the document and a piece of metadata
+    """
+
+    def __init__(self, nodeKey: str, nodeLabel: str, relLabel: str):
+        self.nodeKey = nodeKey
+        self.nodeLabel = nodeLabel
+        self.relLabel = relLabel
+
+
+class GraphExtractor(ABC):
+    def __init__(self, entity_name: str):
+        self._entity_name = entity_name
+
+    @abstractmethod
+    def extract(self, document: "DocSet") -> "DocSet":
+        pass
+
+
+class MetadataExtractor(GraphExtractor):
+    """
+    Extracts metadata from documents and represents them as nodes and relationship in neo4j
+
+    Args:
+        metadata: a list of GraphMetadata that is used to determine what metadata is extracted
+    """
+
+    def __init__(self, metadata: list[GraphMetadata]):
+        self.metadata = metadata
+
+    def extract(self, docset: "DocSet") -> "DocSet":
+        docset.plan = ExtractMetadata(docset.plan, self)
+        docset = self.resolve_metadata(docset)
+
+        return docset
+
+    def extract_metadata(self, doc: Document) -> Document:
+        """
+        Extracts metadata from documents and stores them in the 'nodes' key of 'properties in each document
+        """
+        nodes: Dict[str, Dict[str, Any]] = {}
+        for m in self.metadata:
+            if m.nodeKey not in doc["properties"]:
+                continue
+
+            key = str(m.nodeKey)
+            value = str(doc["properties"][m.nodeKey])
+            node = {"type": "metadata", "properties": {key: value}, "label": str(m.nodeLabel), "relationships": {}}
+            nodes[key + "_" + value] = node
+
+            rel = {
+                "TYPE": str(m.relLabel),
+                "properties": {},
```

Comment:
The doc properties are currently stored in the document node(this is a map function I have that I don't have pushed). I added the properties tag here just in case down the line we wanted to add properties here unsupervised.